lib/datasets/nagoya-university-conversation-corpus.rb

```diff
@@ -0,0 +1,112 @@
+require_relative 'dataset'
+require_relative 'zip-extractor'
+
+module Datasets
+  class NagoyaUniversityConversationCorpus < Dataset
+    Data = Struct.new(
+      :name,
+      :date,
+      :place,
+      :participants,
+      :relationships,
+      :note,
+      :sentences
+    )
+
+    Participant = Struct.new(
+      :id,
+      :attribute,
+      :birthplace,
+      :residence
+    )
+
+    Sentence = Struct.new(
+      :participant_id,
+      :content
+    )
+
+    def initialize
+      super()
+      @metadata.id = 'nagoya-university-conversation-curpus'
+      @metadata.name = 'Nagoya University Conversation Curpus'
+      @metadata.url = 'https://mmsrv.ninjal.ac.jp/nucc/'
+      @metadata.licenses = ['CC-BY-NC-ND-4.0']
+      @metadata.description = <<~DESCRIPTION
+        The "Nagoya University Conversation Corpus" is a corpus of 129 conversations,
+        total about 100 hours of chatting among native speakers of Japanese,
+        which is converted into text.
+      DESCRIPTION
+    end
+
+    def each
+      return to_enum(__method__) unless block_given?
+
+      open_data do |input_stream|
+        yield(parse_file(input_stream))
+      end
+    end
+
+    private
+
+    def open_data
+      data_path = cache_dir_path + 'nucc.zip'
+      data_url = 'https://mmsrv.ninjal.ac.jp/nucc/nucc.zip'
+      download(data_path, data_url)
+
+      zip_file = Zip::File.open(data_path)
+      zip_file.each do |entry|
+        next unless entry.file?
+        ZipExtractor.new(data_path).extract_file(entry.name) do |input_stream|
```

Comment:
Ah, we need to add `ZipExtractor#extract_files` for this use case:

```ruby
class ZipExtractor
  def extract_files
    Zip::File.open(@path) do |zip_file|
      zip_file.each do |entry|
        next unless entry.file?

        entry.get_input_stream do |input|
          yield(input)
        end
      end
    end
  end
end
```

Then we can use one `ZipExtractor` object for this:

```ruby
extractor = ZipExtractor.new(data_path)
extractor.extract_files do |input_stream|
  yield(input_stream)
end
```

Reply:
I will use this method. Thank you.